File: backend/app/schemas/training.py
from __future__ import annotations

from datetime import date, datetime, time
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

TrainingType = Literal["team", "personal", "independent"]
TrainingStatus = Literal["scheduled", "completed", "cancelled"]
AttendanceStatus = Literal["present", "absent"]
ReportStatus = Literal["pending", "accepted", "needs_revision"]


class TrainingCreateIn(BaseModel):
    training_date: date
    start_time: time
    duration_minutes: int = Field(gt=0, le=600)
    location: str | None = Field(default=None, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    plan_id: UUID | None = None
    reminder_minutes_before: int | None = Field(default=None, ge=0, le=10080)
    # If set, also creates a training on the same weekday every week up to
    # and including this date (all sharing one recurrence_group_id).
    repeat_weekly_until: date | None = None
    # Team trainings only: mark this as a self-run session (no coach
    # present) and optionally pick who is responsible for it — defaults to
    # the team captain when omitted.
    is_independent: bool = False
    responsible_user_id: UUID | None = None


class TrainingUpdateIn(BaseModel):
    training_date: date
    start_time: time
    duration_minutes: int = Field(gt=0, le=600)
    location: str | None = Field(default=None, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    plan_id: UUID | None = None
    reminder_minutes_before: int | None = Field(default=None, ge=0, le=10080)
    status: TrainingStatus = "scheduled"
    is_independent: bool = False
    responsible_user_id: UUID | None = None


class TrainingOut(BaseModel):
    id: UUID
    team_id: UUID | None
    created_by: UUID
    type: TrainingType
    plan_id: UUID | None
    training_date: date
    start_time: time
    duration_minutes: int
    location: str | None
    description: str | None
    status: TrainingStatus
    reminder_minutes_before: int | None
    recurrence_group_id: UUID | None
    responsible_user_id: UUID | None
    created_at: datetime
    updated_at: datetime


class AttendanceOut(BaseModel):
    training_id: UUID
    user_id: UUID
    telegram_id: int
    first_name: str
    last_name: str | None
    photo_url: str | None
    status: AttendanceStatus
    marked_by: UUID | None
    marked_at: datetime


class AttendanceStatusIn(BaseModel):
    status: AttendanceStatus

File: frontend/src/types/training.ts
export type TrainingType = "team" | "personal" | "independent";
export type TrainingStatus = "scheduled" | "completed" | "cancelled";
export type AttendanceStatus = "present" | "absent";

export const TRAINING_STATUS_LABELS: Record<TrainingStatus, string> = {
  scheduled: "Запланирована",
  completed: "Проведена",
  cancelled: "Отменена",
};

export function isTrainingOverdue(training: Pick<Training, "status" | "trainingDate">): boolean {
  const today = new Date().toISOString().slice(0, 10);
  return training.status === "scheduled" && training.trainingDate < today;
}

export interface Training {
  id: string;
  teamId: string | null;
  createdBy: string;
  type: TrainingType;
  planId: string | null;
  trainingDate: string;
  startTime: string;
  durationMinutes: number;
  location: string | null;
  description: string | null;
  status: TrainingStatus;
  reminderMinutesBefore: number | null;
  recurrenceGroupId: string | null;
  responsibleUserId: string | null;
}

export interface TrainingCreateInput {
  training_date: string;
  start_time: string;
  duration_minutes: number;
  location: string | null;
  description: string | null;
  plan_id: string | null;
  reminder_minutes_before: number | null;
  repeat_weekly_until: string | null;
  is_independent: boolean;
  responsible_user_id: string | null;
}

export interface TrainingUpdateInput {
  training_date: string;
  start_time: string;
  duration_minutes: number;
  location: string | null;
  description: string | null;
  plan_id: string | null;
  reminder_minutes_before: number | null;
  status: TrainingStatus;
  is_independent: boolean;
  responsible_user_id: string | null;
}

export interface Attendance {
  trainingId: string;
  userId: string;
  telegramId: number;
  firstName: string;
  lastName: string | null;
  photoUrl: string | null;
  status: AttendanceStatus;
}

export interface TrainingDto {
  id: string;
  team_id: string | null;
  created_by: string;
  type: TrainingType;
  plan_id: string | null;
  training_date: string;
  start_time: string;
  duration_minutes: number;
  location: string | null;
  description: string | null;
  status: TrainingStatus;
  reminder_minutes_before: number | null;
  recurrence_group_id: string | null;
  responsible_user_id: string | null;
}

export interface AttendanceDto {
  training_id: string;
  user_id: string;
  telegram_id: number;
  first_name: string;
  last_name: string | null;
  photo_url: string | null;
  status: AttendanceStatus;
}

export function mapTrainingDto(dto: TrainingDto): Training {
  return {
    id: dto.id,
    teamId: dto.team_id,
    createdBy: dto.created_by,
    type: dto.type,
    planId: dto.plan_id,
    trainingDate: dto.training_date,
    startTime: dto.start_time,
    durationMinutes: dto.duration_minutes,
    location: dto.location,
    description: dto.description,
    status: dto.status,
    reminderMinutesBefore: dto.reminder_minutes_before,
    recurrenceGroupId: dto.recurrence_group_id,
    responsibleUserId: dto.responsible_user_id,
  };
}

export function mapAttendanceDto(dto: AttendanceDto): Attendance {
  return {
    trainingId: dto.training_id,
    userId: dto.user_id,
    telegramId: dto.telegram_id,
    firstName: dto.first_name,
    lastName: dto.last_name,
    photoUrl: dto.photo_url,
    status: dto.status,
  };
}

export interface TrainingFeedback {
  trainingId: string;
  userId: string;
  wellbeing: number | null;
  difficulty: number | null;
  comment: string | null;
  skipped: boolean;
}

export interface TrainingFeedbackInput {
  wellbeing: number | null;
  difficulty: number | null;
  comment: string | null;
}

export interface TrainingFeedbackDto {
  training_id: string;
  user_id: string;
  wellbeing: number | null;
  difficulty: number | null;
  comment: string | null;
  skipped: boolean;
}

export function mapTrainingFeedbackDto(dto: TrainingFeedbackDto): TrainingFeedback {
  return {
    trainingId: dto.training_id,
    userId: dto.user_id,
    wellbeing: dto.wellbeing,
    difficulty: dto.difficulty,
    comment: dto.comment,
    skipped: dto.skipped,
  };
}
